text/projects/Lukasz_Selwa/Galil_Seiferas.tex

```diff
@@ -0,0 +1,308 @@
+        \draw (191,35) node [anchor=north west][inner sep=0.75pt]  [font=\small] [align=left] {$v_i$};
+        % Text Node
+        \draw (108,35) node [anchor=north west][inner sep=0.75pt]  [font=\small] [align=left] {$v_j$};
+        % Text Node
+        \draw (60,35) node [anchor=north west][inner sep=0.75pt]  [font=\small] [align=left] {$v_3$};
+        % Text Node
+        \draw (32,35) node [anchor=north west][inner sep=0.75pt]  [font=\small] [align=left] {$v_2$};
+    \end{tikzpicture}
+
+    \caption{Ilustracja do dowodu lematu.}
+    \label{fig:Lem_3_proof}
+\end{figure}
+
+Należy rozważyć dwa przypadki. Pierwszy kiedy $v_j$ jest prefiksem $z$. Wtedy za pomocą lematu \ref{Lem_3} można pokazać, że $w$ nie jest słowem pierwotnym.
+Drugi przypadek jest kiedy początek $z y$ znajduje się ściśle wewnątrz $v_j$. 
+Za pomocą lematu \ref{Lem_3} można pokazać, że w tym przypadku $v_j$ nie może być słowem pierwotnym.
+
+\begin{definition}
+\textbf{\textit{Pozycja specjalna.}} Lemat \ref{Lem_Key} pozwala na zdefiniowanie pojęcia pozycji specjalnych. Niech $V(x) = (v_1, ...)$ to ciąg czynników $x$. Pierwszą pozycją specjalną słowa $x$ jest długość słowa $v_1...v_i$ takiego, że $v_{i+1}$ albo nie istnieje albo $|v_1...v_{i+1}| \geq \HRP2(x)$. Jeśli $x = v_1...v_i x'$ to kolejną pozycją specjalną definiujemy analogicznie dla $x'$. Robimy tak dopóki dla $x$ istnieje $\HRP2(x)$.
+\end{definition}
+
+\begin{theorem}[O dekompozycji]\label{Th_Decomposition}
+Niech $j$ będzie ostatnią specjalną pozycją słowa $x$.
+Niech $u = x[1...j], v=x[j+1...n]$. Wtedy $uv$ jest $k$-perfekcyjną dekompozycją $x$.
+\end{theorem}
+Z definicji $v$ nie może mieć $\HRP2(v)$ pozostało zauważyć, że $|u| < p(v)$.
+Kluczową obserwacją jest fakt, że dla kolejnych pozycji specjalnych $j$, $j'$ zachodzi $2\cdot |\HRP2(x[j...])| \leq |\HRP2(x[j'...])|$. Wynika to z lematów \ref{Lem_3}. \ref{Lem_Key}.
+Na podstawie twierdzenia \ref{Th_Decomposition} możemy skonstruować algorytm do wyznaczania $k$-perfekcyjnej dekompozycji.
+\begin{algorithm}[H]
+\caption{PERFECT-DECOMPOSITION}
+\label{perfect_decomposition}
+\begin{algorithmic} 
+\STATE $j \gets 0, hrp1 \gets |\HRP1(x)|, hrp2 \gets |\HRP2(x)|$
+\WHILE{$hrp1 \neq \texttt{NULL} \land hrp2 \neq \texttt{NULL}$}
+    \STATE $j \gets j + hrp1$
+    \STATE $hrp1 \gets |\HRP1(x[j...n])$
+    \IF{$hrp1 \geq hrp2$}
+        \STATE $hrp2 \gets|\HRP2(x[j...n])$
+    \ENDIF
+\ENDWHILE
+\end{algorithmic}
+\end{algorithm}
+
+Galil Seiferas pokazali, że jeśli potrafimy obliczyć $\HRP1(x)$ oraz $\HRP2(x)$ w czasie proporcjonalnym do $|\HRP2(x)|$ to powyższy algorytm jest liniowy. Dowód tego faktu pominiemy.
+Pokażemy za to sam algorytm wyznaczania $\HRP1(x)$ i $\HRP2(x)$ w odpowiednim czasie.
+
+\subsection{Wyznaczanie $k$-HRP}
+
+Przedstawimy jak wyznaczyć wszystkie $\HRP$ słowa $w$ o długości $n$.
+Załóżmy, że mamy policzony ciąg pierwszych $\HRP$ $v_1, v_2, ..., v_r$ i ich odpowiednie zakresy $[L_1, R_1], [L_2, R_2], ..., [L_r, R_r]$.
```

Comment:
This is unclear to me: is this a "sequence of factors" as described by Def. 2.4? And how exacly do you square it with Lemma 2.1 so that $[L_r, R_r]$ and $[L_{r + 1}, R_{r + 1}]$ don't overlap?